pkg/api/config.go

```diff
@@ -1,27 +1,199 @@
+
+	// Validate BuildRootImageConfiguration
+	case buildRootImageConfigurationType:
+		validationErrors = append(validationErrors,
+			validateBuildRootImageConfiguration(fieldRoot, iface.Interface().(BuildRootImageConfiguration)))
+
+	// Validate ImageStreamTagReference
+	case imageStreamTagReferenceType:
+		validationErrors = append(validationErrors,
+			validateImageStreamTagReference(fieldRoot, iface.Interface().(ImageStreamTagReference)))
+	}
+
+	if iface.Kind() == reflect.Struct {
+		for i := 0; i < iface.NumField(); i++ {
+			f := iface.Field(i)
+			fieldRoot = append(fieldRoot, iface.Type().Field(i).Tag)
+
+			// Resolve pointer's value.
+			if f.Kind() == reflect.Ptr {
+				f = reflect.Indirect(f)
 			}
+
+			switch f.Kind() {
+			case reflect.Struct:
+				// Recursion
+				validationErrors = append(validationErrors, deepValidate(f, fieldRoot))
+
+			case reflect.Slice:
+				// Validate []TestStepConfiguration
+				for i := 0; i < f.Len(); i++ {
+					if f.Index(i).Type() == testStepConfigurationType {
+						validationErrors = append(validationErrors, validateTestStepConfiguration(fieldRoot, f.Index(i).Interface().(TestStepConfiguration)))
+					}
+				}
+
+			case reflect.Map:
+				// Validate map[string]ImageStreamTagReference
+				if f.Type() == imageStreamTagReferenceMapType {
+					imageStreamTagReferenceMap := f.Interface().(map[string]ImageStreamTagReference)
+
+					// map has to be sorted to avoid flakes in tests.
+					var keys []string
+					for k := range imageStreamTagReferenceMap {
+						keys = append(keys, k)
+					}
+					sort.Strings(keys)
+
+					for _, k := range keys {
+						validationErrors = append(validationErrors,
+							validateImageStreamTagReference(
+								append(fieldRoot, reflect.StructTag(fmt.Sprintf("json:\"%s\"", k))), imageStreamTagReferenceMap[k]))
+					}
+				}
+			}
+
+			// Clear fieldRoot Tags
+			fieldRoot = []reflect.StructTag{}
 		}
-	} else if buildRootImage != nil && testBaseImage != nil {
-		return errors.New("both build_root and test_base_image cannot be set")
-	} else if buildRootImage != nil && testBaseImage == nil {
-		if buildRootImage.ProjectImageBuild != nil && buildRootImage.ImageStreamTagReference != nil {
-			return errors.New("both git_source_image and image_stream_tag cannot be set for the build_root")
-		} else if buildRootImage.ProjectImageBuild == nil && buildRootImage.ImageStreamTagReference == nil {
-			return errors.New("you have to specify either git_source_image or image_stream_tag for the build_root")
-		}
-	} else if buildRootImage == nil && testBaseImage == nil {
-		return errors.New("no build_root or test_base_image has been set")
+	}
+	return kerrors.NewAggregate(validationErrors)
+}
+
+// Validate validates all the configuration's values.
+func (config *ReleaseBuildConfiguration) Validate() error {
+	if err := deepValidate(reflect.ValueOf(config), []reflect.StructTag{}); err != nil {
+		return err
 	}
 	return nil
 }
```

Comment:
The individual struct validation functions below look good but an `if`-based approach would be more straight-forward in this case than using reflection, as @coverprice mentioned.

This is [how kubernetes does validation](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go).